From c9a31bfd3fc3a155e88c40934ea257b528a04f6a Mon Sep 17 00:00:00 2001
From: Russell Webb <russell.webb@intel.com>
Date: Thu, 14 Nov 2013 19:07:24 +0100
Subject: Fix problem with BootAnimation exit

After the completion of each animation part, the boot animation must
sleep for the number of frames specified in part.pause.  If this is
done in a single sleep statement, it is possible that the boot animation
will continue to display long after surface flinger has finished booting.

To address that, the pause between parts is done in a loop and on each
iteration, the property that surface flinger sets to indicate that animation
should exit is checked.  This minimizes, but doesn't remove, the likelihood
that the bootanimation is visible when the UI comes up.

Category: AOSP Enhancement
Upstream Candidate: Yes
Domain: AOSP
For: APDEV-53
Change-Id: I93b1aef8e04ffc9269a471c2cfaac17f590d3a07
Signed-off-by: Dariusz Szymczak <dariusz.szymczak@intel.com>
---
 cmds/bootanimation/BootAnimation.cpp |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/cmds/bootanimation/BootAnimation.cpp b/cmds/bootanimation/BootAnimation.cpp
index ad4e4c8..3fe7e76 100644
--- a/cmds/bootanimation/BootAnimation.cpp
+++ b/cmds/bootanimation/BootAnimation.cpp
@@ -555,7 +555,16 @@ bool BootAnimation::movie()
                 checkExit();
             }
 
-            usleep(part.pause * ns2us(frameDuration));
+            // part.pause is the number of frames to pause for so total sleep will be
+            // part.pause * frameDuration.  Instead of a single sleep call, sleep for
+            // frameDuration and then check if surface flinger is done.
+            const nsecs_t frameDurationUs = ns2us(frameDuration);
+            for (int k = 0; k < part.pause; k++) {
+                usleep(frameDurationUs);
+                checkExit();
+                if(exitPending())
+                    break;
+            }
 
             // For infinite parts, we've now played them at least once, so perhaps exit
             if(exitPending() && !part.count)
-- 
1.7.4.1

